Support Chained Transactions

h0ngcha0 opened the pull request:
Support chained transactions.

1. Support submitting chained transactions
2. If the connected account is performing a dApp transaction (execute script or deploy contract) but does not have enough balance, it will try to transfer from an account with enough balance even if the account is in another group
3. Ledger account doesn't support chained tx yet

![image](https://github.com/user-attachments/assets/65a8b06a-ab08-45ac-bfee-0a935092e88b)
![image](https://github.com/user-attachments/assets/0debef9a-7b63-4821-bd72-6567eab9dafc)

Lbqds commented on packages/extension/src/ui/features/actions/ActionScreen.tsx:
```diff
@@ -32,7 +32,7 @@ export const ActionScreen: FC = () => {
   const isLastAction = actions.length === 1
   const signerAccount = useAccount(action.type === 'ALPH_TRANSACTION' && selectedAccount
-    ? { address: action.payload.params.signerAddress, networkId: action.payload.params.networkId ?? selectedAccount.networkId }
+    ? { address: action.payload[0].params.signerAddress, networkId: action.payload[0].params.networkId ?? selectedAccount.networkId }  // FIXME: always use the first payload
```
Do we need to fix this? It seems like it doesn't matter since we always sign the transaction with the specified address in `TransactionParams`.

Lbqds commented on packages/extension/src/ui/features/actions/ActionScreen.tsx:
```diff
@@ -134,20 +134,30 @@ export const ActionScreen: FC = () => {
+                navigate(routes.error())
+              }
+
+              await approveAction(action, builtTransactions)
```
Shall we wrap these lines of code in an `else` branch?

h0ngcha0 replied:
It should be the same without the `else` branch since it would navigate to the error page, but perhaps more clear with the `else` statement. Will fix.

Lbqds commented on packages/extension/src/shared/transactions/index.ts:
```diff
@@ -129,3 +133,295 @@ function buildGetTransactionsFn(metadataTransactions: Transaction[]) {
+  transactionParams: TransactionParams
+): Promise<Map<string, BigNumber> | undefined> {
+  const expectedBalances: Map<string, BigNumber> = new Map()
+  const gasFee = BigInt(5000000) * DEFAULT_GAS_PRICE  // Maximal gas fee per tx
```
If the user specifies both `gasAmount` and `gasPrice`, maybe we could use `gasAmount * gasPrice`. WDYT?

Lbqds commented on packages/extension/src/shared/transactions/index.ts:
```diff
@@ -129,3 +133,295 @@ function buildGetTransactionsFn(metadataTransactions: Transaction[]) {
+      addTokenToBalances(expectedBalances, ALPH_TOKEN_ID,
+        transactionParams.params.initialAttoAlphAmount !== undefined
+          ? BigNumber.from(transactionParams.params.initialAttoAlphAmount)
+          : BigNumber.from(ONE_ALPH)
```
[Nit]: we can use the `MINIMAL_CONTRACT_DEPOSIT` constant from the SDK.

Lbqds commented on packages/extension/src/shared/transactions/index.ts:
```diff
@@ -129,3 +133,295 @@ function buildGetTransactionsFn(metadataTransactions: Transaction[]) {
+      }
+
+      const [firstMissingTokenId, firstMissingAmount] = missingBalances.entries().next().value;
+      const tokenSymbol = firstMissingTokenId === ALPH_TOKEN_ID ? 'ALPH' : firstMissingTokenId;
```
Is it possible to get the token symbol based on the token id?

h0ngcha0 replied:
Addressed now, thanks :+1:

Lbqds commented on packages/extension/src/inpage/alephiumWindowObject.ts:
```diff
@@ -205,8 +211,96 @@ export const alephiumWindowObject: AlephiumWindowObject =
+        throw Error("Empty transaction params")
+      }
+
+      const transactionParamz: TransactionParams[] = params.map(param => {
```
[Nit]: it seems that we can use the `signedChainedTxParamsToTransactionParams` function.

h0ngcha0 replied:
I initially used `signedChainedTxParamsToTransactionParams` and when interacting with the dApp it complained that this function can only be used by the extension (due to where its defined). I will see if there is a way to fix it.

h0ngcha0 replied:
Fixed now, put it in the [transformer.ts](https://github.com/alephium/extension-wallet/blob/chained-txs/packages/extension/src/shared/transactions/transformers.ts#L21) instead of [index.ts](https://github.com/alephium/extension-wallet/blob/chained-txs/packages/extension/src/shared/transactions/index.ts) works.

Lbqds reviewed (approved):
LGTM 👍

polarker commented:
@h0ngcha0 Could you please update the screenshot as well, since I won't have the time to run it by myself?

h0ngcha0 commented:
> @h0ngcha0 Could you please update the screenshot as well, since I won't have the time to run it by myself?

Definitely, the the screenshots below shows when account doesn't have enough ALPH to deploy a contract and enough ALPH is transferred to this account first:

![image](https://github.com/user-attachments/assets/f1be2792-c958-4e09-9118-f99aa47e9d60)  
![image](https://github.com/user-attachments/assets/a84b3f13-33fb-43c2-aff0-3091b191d15b)

polarker commented:
@h0ngcha0 Maybe we should add "< 1 of 2 >" at the top of the panel, so it's not hidden when the transaction details are long.

polarker commented on packages/extension/src/background/actionHandlers.ts:
```diff
@@ -49,21 +50,61 @@ export const handleActionApproval = async (
+          const { signatures } = await executeTransactionsAction(
+            transactions,
+            background,
+            transactions[0].params.networkId,
```
Let's verify that all network IDs are the same before using the first one. Checking ahead of time can help prevent potential bugs, even though we'll also confirm in the Approval phase.

polarker commented on packages/extension/src/background/transactions/transactionExecution.ts:
```diff
@@ -29,15 +29,45 @@ export const executeTransactionAction = async (
+      signerAddress: account.address,
+      unsignedTx: transaction.result.unsignedTx,
+    })
+
```
Why is the signing code logic different from executeTransactionAction?

h0ngcha0 replied:
Great question. I have tested that the `signatureOpt` in `executeTransactionAction` is not needed. I will remove it so it is consistent with `executeTransactionsAction`.

polarker reviewed (approved):
LGTM, just a few minor points. Didn't have time to review the chained transaction logic line by line, but that's okay since we're migrating to a full node anyway.

h0ngcha0 commented:
> @h0ngcha0 Maybe we should add "< 1 of 2 >" at the top of the panel, so it's not hidden when the transaction details are long.

Moved the transaction counter on the top like this, also removed the `<` and `>` buttons since they duplicate with the buttons on the bottom of the page. WDYT?

![image](https://github.com/user-attachments/assets/d194a77a-04ee-4854-8699-179c3363170d)
![image](https://github.com/user-attachments/assets/b1e30222-cd39-42c4-9a9a-66ef3702c596)

polarker commented:
That UI change looks great to me!
